### ISSUE TITLE
Narrowing via property check fails when adding parent type to declaration

### ISSUE THREAD

**Issue by danvk** [microsoft/TypeScript#37518]
**TypeScript Version:**  3.8.3

**Search Terms:**
- property check refinement

**Expected behavior:** No error, type is narrowed to `SubA`.

**Actual behavior:** Type is narrowed to `never` and produces an error.

**Related Issues:**
- https://github.com/microsoft/TypeScript/issues/1260

**Code**

This code type checks:

```ts
interface Base {
  base: number;
}

interface SubA extends Base {
  a: string;
  other: boolean;
}
interface SubB extends Base {
  b: string;
}

function foo(x: SubA | SubB | null) {
  x = x || {base: 1, b: 'b'};
  if (x && 'a' in x && x.a == 'str') {
    x.other;  // ok
  }
}
```

A very tiny tweak (adding `Base |` to the parameter type) breaks things, though:

```ts
function foo(x: Base | SubA | SubB | null) {
  x = x || {base: 1, b: 'b'};
  if (x && 'a' in x && x.a == 'str') {  // <-- Property 'a' does not exist on type 'never'. (2339)
    x.other;  // <-- Property 'other' does not exist on type 'never'. (2339)
  }
}
```

This makes some sense since `foo` could be called with something assignable to `Base` that has an `a` but is not specifically a `SubA`. But what doesn't make sense to me is that removing the assignment on the first line makes the error go away:

```ts
function foo(x: Base | SubA | SubB | null) {
  if (x && 'a' in x && x.a == 'str') {  // ok
    x.other;  // ok
  }
}
```

That assignment seeming has no bearing on `if` statement (the default value doesn't have an `a` property). So why does it affect the inferred type in it?

Either the second and third example should both error or neither of them should.

<details><summary><b>Output</b></summary>

```ts
"use strict";
function foo(x) {
    x = x || {base: 1, b: 'b'};
    if (x && 'a' in x && x.a == 'str') {
        x.other;
    }
}

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "moduleResolution": 2,
    "target": "ES2017",
    "jsx": "React",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgEJwM4oN4ChnIBGmEAXMiAK4C2h0A3LgL666iSyIoDKlhAgsggAPSCAAmGNCWR4CcchjBRQAc0YEA9mAAW0coU2aANhDghGLdtHhJkvQqiGiIEqeiyz8RRcrWXWGEoQBDBgTRBkGCMACmFyDxQAH3s+QRSHJxSqY2MASi8CYWQAXmRipJTsYixyAEYAGh9kAHJCFqYNZGAYZDjkADIB1rgW7sjiofKAOjhSspalKBaCuQIZ7T0oLpYmIA)

**Comment by jack-williams** [comment]
> But what doesn't make sense to me is that removing the assignment on the first line makes the error go away

The `||` operator triggers subtype reduction, so in the second example the narrowed identifier has type `Base`, while in the third example the identifier has type `Base | SubA | SubB | null`.

The `in` operator narrows using the assumption that types are closed, so `Base | SubA | SubB` noes not behave the same as `Base`, despite the types being semantically the same.

Removing the closed assumption from `in` would make it useless, so I think the only way to have consistent behaviour is apply subtype reduction prior to `in`.
